strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java

```diff
@@ -106,57 +116,69 @@ protected RepositoryPath resolvePathTraversal(RepositoryPath repositoryPath) thr
             {
                 continue;
             }
-            
-            RepositoryPath result = repositoryPathResolver.resolve(r, repositoryPath);
-            if (artifactRoutingRulesChecker.isDenied(groupRepository.getId(), result))
+
+            final RepositoryPath resolvedPath = repositoryPathResolver.resolve(r, repositoryPath);
+            if (artifactRoutingRulesChecker.isDenied(groupRepository.getId(), resolvedPath))
             {
                 continue;
             }
-            
-            result = resolvePathFromGroupMemberOrTraverse(result);
-            if (result == null)
+
+            if (!firstMatch)
             {
+                aggregatedActions.add(() -> resolvePathFromGroupMemberOrTraverse(resolvedPath, false));
                 continue;
             }
-            
-            logger.debug(String.format("Located artifact: [%s]", result));
-            
-            return result;
+
+            final RepositoryPath groupMemberPath = resolvePathFromGroupMemberOrTraverse(resolvedPath, true);
+            if (groupMemberPath != null)
+            {
+                logger.debug(String.format("Located artifact: [%s]", groupMemberPath));
+                return groupMemberPath;
+            }
+        }
+
+        if (!firstMatch)
+        {
+            invokeResolvePathActions(aggregatedActions);
+            return repositoryPath;
         }
+
         return null;
     }
 
+    private void invokeResolvePathActions(List<Callable<RepositoryPath>> resolvePathActions)
+    {
+        try
+        {
+            // TODO invokeAll timeout interrupt ?
+            executorService.invokeAll(resolvePathActions);
+        }
+        catch (InterruptedException e)
+        {
+            Thread.currentThread().interrupt();
```

Comment:
why this needed? 
as I understand `InterruptedException` occurs when thread already been interrupted :)

Reply:
https://www.ibm.com/developerworks/library/j-jtp05236/index.html